sharktank/tests/models/llama/sharded_llama_test.py

```diff
@@ -5,122 +5,119 @@
+            "attention_mask": attention_mask,
+            "seq_block_ids": seq_block_ids,
+            "cache_state": cache_state,
+        }
 
-        # Verify prefill step.
-        sharded_config = deepcopy(config)
-        sharded_config.tensor_parallelism_size = 2
-        sharded_theta = shard_theta(theta, sharded_config)
-        sharded_model = PagedLlamaModelV1(sharded_theta, sharded_config)
+    def make_equal_unsharded_and_sharded_prefill_args(
+        self, model: PagedLlamaModelV1, sharded_model: PagedLlamaModelV1
+    ) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+        prefill_args = self.make_prefill_args(model)
         sharded_cache_state = sharded_model.cache.paged.allocate(
-            page_count=cache_page_count
+            page_count=self.cache_page_count
+        )
+        assert iterables_equal(
+            prefill_args["cache_state"][0].shape, sharded_cache_state[0].shape
         )
+        sharded_prefill_args = deepcopy(prefill_args)
         sharded_cache_state = sharded_model.cache.paged.shard_state(
-            deepcopy(cache_state)
+            sharded_prefill_args["cache_state"]
         )
+        sharded_prefill_args["cache_state"] = sharded_cache_state
+        return prefill_args, sharded_prefill_args
 
-        expected_prefill_result = model.prefill(
-            token_ids,
-            attention_mask=attention_mask,
-            seq_block_ids=seq_block_ids,
-            cache_state=cache_state,
-        )
-        sharded_prefill_result = sharded_model.prefill(
-            token_ids,
-            attention_mask=attention_mask,
-            seq_block_ids=seq_block_ids,
-            cache_state=sharded_cache_state,
-        )
-        # The errors are quite high, but for float64 both errors drop to < 1e-12.
-        # The numerics are probably correct.
-        torch.testing.assert_close(
-            sharded_prefill_result, expected_prefill_result, atol=1e-3, rtol=1e-2
+    def make_decode_args(self, model: PagedLlamaModelV1) -> Dict[str, Any]:
+        seq_lens = torch.randint(
+            high=self.config.hp.context_length + 1, size=[self.batch_size]
         )
-        expected_cache_state = cache_state[0]
-        actual_cache_state = ops.unshard(
-            sharded_model.cache.paged.unflatten_page_table(sharded_cache_state)
-        ).flatten(start_dim=1)
-        torch.testing.assert_close(
-            actual_cache_state, expected_cache_state, atol=1e-4, rtol=1e-1
+        seq_lens[self.batch_size - 1] = self.config.hp.context_length
+        batch_seq_len = round_up_to_multiple_of(
+            int(torch.max(seq_lens)), model.cache.pad_sequence_stride
         )
-
-        # Verify decode step.
         decode_token_ids = torch.randint(
             low=0,
-            high=vocabulary_size,
-            size=[batch_size, 1],
+            high=self.vocabulary_size,
+            size=[self.batch_size, 1],
             dtype=torch.int32,
         )
         decode_seq_lens = torch.randint(
```

Comment:
Same about above. We would expect the decode to be one more than the sequence length of the encode. We should maintain that behavior.

Reply:
About the sequence length I think that is the case, but the cache is being regenerated again. In this test it really does not matter if we get plausible numbers, just that the numbers are close.

Reply:
I made them `prefill_seq_lens + 1`.